File: code/src/main/java/encapsulation/FPUsage.java
package encapsulation;

import java.net.InetAddress;
import java.util.function.Consumer;

import static encapsulation.EmailAddress.parse;
import static encapsulation.EncapsulationKt.inetAddress;
import static encapsulation.EncapsulationKt.sendEmail;
import static encapsulation.FPUsageLambda.createEmailSender;

public class FPUsage {
    /// begin: foo
    // Where we know the configuration
    Consumer<Email> sender = createEmailSender(
        inetAddress("example.com"),
        "username",
        "password"
    );

    // Where we send the message
    public void sendThanks() {
        sender.accept( // <1>
            new Email(
                parse("support@internationalrescue.org"),
                parse("support@travelator.com"),
                "Thanks for your help",
                "..."
            )
        );
    }
    /// end: foo
}

class FPUsageLongHand {
    /// begin: FPUsageLongHand
    static Consumer<Email> createEmailSender(
        InetAddress serverAddress,
        String username,
        String password
    ) {
        return new Consumer<Email>() {
            @Override
            public void accept(Email email) {
                sendEmail(
                    email,
                    serverAddress,
                    username,
                    password
                );
            }
        };
    }
    /// end: FPUsageLongHand
}

class FPUsageLambda {
    /// begin: FPUsageLambda
    static Consumer<Email> createEmailSender(
        InetAddress serverAddress,
        String username,
        String password
    ) {
        return email -> sendEmail(
            email,
            serverAddress,
            username,
            password
        );
    }
    /// end: FPUsageLambda
}

File: code/src/main/java/encapsulation/FP-usage.kt
package encapsulation

import encapsulation.EmailAddress.Companion.parse
import encapsulation.EmailSenderInvoke.EmailSender

object FP {
    /// begin: foo
    // Where we know the configuration
    val sender: (Email) -> Unit = createEmailSender(
        inetAddress("smtp.travelator.com"),
        "username",
        "password"
    )

    // Where we send the message
    fun sendThanks() {
        sender( // <1>
            Email(
                to = parse("support@internationalrescue.org"),
                from = parse("support@travelator.com"),
                subject = "Thanks for your help",
                body = "..."
            )
        )
    }
    /// end: foo
}

object FP_invoke {
    // Where we know the configuration
    val sender: (Email) -> Unit = createEmailSender(
        inetAddress("smtp.travelator.com"),
        "username",
        "password"
    )

    // Where we send the message
    /// begin: nonInvoke
    fun sendThanks() {
        sender(
            Email(
                to = parse("support@internationalrescue.org"),
                from = parse("support@travelator.com"),
                subject = "Thanks for your help",
                body = "..."
            )
        )
    }
    /// end: nonInvoke
}

object FP_OO {
    /// begin: bar
    // Where we know the configuration
    val sender: (Email) -> Unit = EmailSender(
        inetAddress("smtp.travelator.com"),
        "username",
        "password"
    )

    // Where we send the message
    fun sendThanks() {
        sender( // <1>
            Email(
                to = parse("support@internationalrescue.org"),
                from = parse("support@travelator.com"),
                subject = "Thanks for your help",
                body = "..."
            )
        )
    }
    /// end: bar
}
